File: backend/app/dto/review.py
from typing import Optional
from app.dto.movie import MovieGetResponse
from app.dto.achievement import AchievementDTO
from pydantic import BaseModel, Field
from datetime import datetime as datetime
from fastapi_pagination import Page

class ReviewDTO(BaseModel):
    id: int
    user_id: int
    movie_id: int
    rating: Optional[int] = Field(None, ge=0, le=5)
    text: Optional[str] = None
    watch_date: datetime
    likes: int
    liked_by_user: bool = False
    created_at: datetime
    movie: Optional[MovieGetResponse] = None

class ReviewCreationDTO(BaseModel):
    movie_id: int
    rating: Optional[int] = Field(None, ge=0, le=5)
    text: Optional[str] = None
    watch_date: Optional[datetime] = None

class ReviewGetSingularDTO(ReviewDTO):
    user_name: str

class ReviewGetSingularAchievementsDTO(ReviewGetSingularDTO):
    achievements: list[AchievementDTO] = []

class ReviewGetResponse(BaseModel):
    user_review: Optional[ReviewGetSingularAchievementsDTO] = None
    reviews: Page[ReviewGetSingularAchievementsDTO]

File: frontend/app-front/services/api/flixy/types/review.ts
import type { CommentDataGet } from "./comment";
import type { MovieDataGet } from "./movie";
import type { Page } from "./overall";
import type { UserAchievement } from "./user";

export interface ReviewCreation {
  movie_id: number;
  rating?: number;
  text?: string;
  watch_date?: Date;
  [key: string]: string | number | Date | undefined;
}

export interface ReviewDataGet {
  id: number;
  name: string;
  user_name: string;
  user_id: number;
  movie_id: number;
  rating: number | null;
  text: string | null;
  watch_date: Date;
  likes: number;
  liked_by_user: boolean;
  updated_at: Date;
  movie: MovieDataGet;
  achievements: UserAchievement[];
  comments: CommentDataGet[];
  [key: string]:
    | string
    | number
    | Date
    | boolean
    | MovieDataGet
    | UserAchievement[]
    | CommentDataGet[]
    | null
    | undefined;
}

export interface ReviewsData {
  user_review?: ReviewDataGet | undefined;
  reviews?: Page<ReviewDataGet> | undefined;
  [key: string]: ReviewDataGet | Page<ReviewDataGet> | undefined;
}
